src/worker/mod.rs

```diff
@@ -369,45 +394,83 @@ impl<
                 .await?;
         }
 
-        let exit = self.listen_for_heartbeats(&statuses).await;
-
-        // there are a couple of cases here:
-        //
-        //  - we got TERMINATE, so we should just return, even if a worker is still running
-        //  - we got TERMINATE and all workers have exited
-        //  - we got an error from heartbeat()
-        //
-        self.terminated = exit.is_ok();
-
-        if let Ok(true) = exit {
-            let running = self.force_fail_all_workers().await;
-            if running != 0 {
-                return Ok(running);
-            }
-        }
+        // for "forever" operations (currently only `Worker::run_to_completion`) we support SIGTERM handling,
+        // whereas for long-running operations (see `Worker::run`), we are polling the cancellation future which
+        // serves as a signal to start graceful shutdowna and return control to the calling site
+        let cancel_signal = if self.forever {
+            None
+        } else {
+            self.shutdown_signal.take()
+        };
 
-        // we want to expose worker errors, or otherwise the heartbeat error
-        let mut results = Vec::with_capacity(workers_count);
-        while let Some(res) = join_set.join_next().await {
-            results.push(res.expect("joined ok"));
-        }
+        let report = tokio::select! {
+            // A signal SIGTERM from the OS received.
+            _ = tokio::signal::ctrl_c(), if self.forever => {
```

Comment:
I still don't want us to have any signal handling code inside the core library. Instead, we should make it possible to configure `run_to_completion` with a graceful shutdown feature as well.

Reply:
Removed this select arm and also the tokio's `signal` feature from the prod deps.

I guess the reason why we are deciding against ctrl_c handling is the possible conflicts from the user space code event handling. They should (at least will want to) define it in their binary.